# Exercises SW04

## Exercise 5: Probability for a collision
<b>There are 40 people in a room. You bet, that there are at least 2 people with the same birthday. What is
the probability, that You win? Use the exact formula as well as the approximation!</b>

In [13]:
# Working with the exact formula
from sympy import *
i = symbols('i',integer=True)
p = 1 - product((365-i)/365,(i,1,39))
p.evalf()

0.891231809817949

Die Wahrscheinlichkeit, dass 2 Personen denselben Geburtstag haben, beläuft sich auf ca. <b>89.123%</b>

## Exercise 6: Probability of a collision 
<b>Suppose You use a hash function of length 128 bits. How many hash values would You have to compute
in order to find a collision with probability at least 90%? where m = 128 and p = 0:9. How much storage would You need, if You need 16 bytes for each hash?</b>

In [10]:
import numpy as np
n = (2 ** ((128+1)/2)) * np.sqrt([np.log((1/(1-0.9)))])
print(n)

[3.95860861e+19]


3.95860861e+19 = 3.9586 * 10^19 bits

3.9586 * 10^19 bits = 4.94825 * 10^18 bytes

4.94825 * 10^18 bytes = <b>4'948'250 terabytes</b>

4'948'250 terabytes * 128 = <b>633'376'000 terabytes</b>